mxclient/public-rooms.go

```diff
@@ -71,12 +72,29 @@ func (r *WorldReadableRooms) Update() error {
 	return nil
 }
 
-// For future when we support filtering the public room directory (LOCALLY)
-//func (r *WorldReadableRooms) GetFilteredPage(page, pageSize int, query string) []gomatrix.PublicRoom {
-//	r.roomsMutex.RLock()
-//	defer r.roomsMutex.RUnlock()
-//	return nil
-//}
+// GetFilteredPage returns a filtered & paginated slice of the WorldReadableRooms Collection
+func (r *WorldReadableRooms) GetFilteredPage(page, pageSize int, query string) []gomatrix.PublicRoom {
+	r.roomsMutex.RLock()
+	defer r.roomsMutex.RUnlock()
+
+	lowerQuery := strings.ToLower(query)
+
+	filteredRooms := make([]gomatrix.PublicRoom, 0, pageSize)
+	for _, room := range r.rooms {
+		if len(filteredRooms) > pageSize {
+			break
+		}
+
+		if (lowerQuery[0] == '#' && strings.Contains(strings.ToLower(room.CanonicalAlias), lowerQuery)) ||
```

Comment:
This will panic if `lowerQuery` is a zero-length string, so best to check `len(lowerQuery) > 0` somewhere.

Reply:
Are you sure?


```
// Index returns the index of the first instance of substr in s, or -1 if substr is not present in s.
func Index(s, substr string) int {
	n := len(substr)
	switch {
	case n == 0:
		return 0
```

looks to have an explicit check for the 0-len substr case

Inspecting the strings stdlib only `Repeat` seems to be panicky

Reply:
I was referring to the `lowerQuery[0] == '#'` bit :-)

Reply:
oh 🤦